File: practice.py
import csv

class Cryptocurrency:
    def __init__(self, name: str, symbol: str, price: float, market_cap: float, circulating_supply: float):
        self.name = name
        self.symbol = symbol
        self.price = price
        self.market_cap = market_cap
        self.circulating_supply = circulating_supply

def clean_string(value: str) -> str:
    return value.replace('"', '').replace(',', '.')

def load_cryptos(filename: str) -> list[Cryptocurrency]:
    cryptos = []
    try:
        with open(filename, 'r', encoding='utf-8') as file:
            reader = csv.reader(file)
            next(reader)  # Пропускаем заголовок
            
            for row in reader:
                try:
                    cryptos.append(Cryptocurrency(
                        name=clean_string(row[0]),
                        symbol=clean_string(row[1]),
                        price=float(clean_string(row[2])),
                        market_cap=float(clean_string(row[3])),
                        circulating_supply=float(clean_string(row[4]))
                    ))
                except (IndexError, ValueError) as e:
                    print(f"Ошибка в строке: {row} - {e}")
    except FileNotFoundError:
        print("Файл не найден!")
    return cryptos

def print_crypto(crypto: Cryptocurrency) -> None:
    print(f"\nНазвание: {crypto.name}")
    print(f"Символ: {crypto.symbol}")
    print(f"Цена: {crypto.price:.2f} USD")
    print(f"Капитализация: {crypto.market_cap:.2f}")
    print(f"Оборот: {crypto.circulating_supply:.2f}")

def find_crypto(cryptos: list[Cryptocurrency], name: str) -> Cryptocurrency | None:
    for crypto in cryptos:
        if crypto.name.lower() == name.lower():
            return crypto
    return None

def main():
    cryptos = load_cryptos('currencies25.csv')
    if not cryptos:
        return

    while True:
        print("\n1. Поиск по названию")
        print("2. Показать все")
        print("0. Выход")
        choice = input("Выберите: ").strip()

        if choice == "0":
            break
        elif choice == "1":
            query = input("Название: ").strip()
            found = find_crypto(cryptos, query)
            print_crypto(found) if found else print("Не найдено!")
        elif choice == "2":
            for crypto in cryptos:
                print_crypto(crypto)
        else:
            print("Неверный ввод!")

if __name__ == "__main__":
    main()

File: practice.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <iomanip>
#include <algorithm>

const int MAX_CRYPTOS = 100;

using namespace std;

struct Cryptocurrency
{
    string Name;
    string Symbol;
    double Price;
    double Market_cap;
    double Circulating_supply;
};

struct ListNode
{
    Cryptocurrency data;
    ListNode *next;
};

struct LinkedList
{
    ListNode *head = nullptr;
    ListNode *tail = nullptr;
};

string replace_comma_with_dot(const string &input)
{
    string result = input;
    replace(result.begin(), result.end(), ',', '.');
    return result;
}

string remove_quotes(const string &input)
{
    string result = input;
    result.erase(remove(result.begin(), result.end(), '\"'), result.end());
    return result;
}

void load_from_file(LinkedList &list)
{
    ifstream file("currencies25.csv");
    if (!file.is_open())
    {
        cout << "Ошибка в открытии файла!" << endl;
        return;
    }

    string line;
    // Пропускаем только строку заголовка
    if (!getline(file, line))
    {
        cerr << "Ошибка: файл пуст!" << endl;
        return;
    }

    while (getline(file, line))
    {
        if (line.empty())
            continue;

        stringstream ss(line);
        Cryptocurrency crypto;

        try
        {
            // Name
            string field;
            getline(ss, field, ',');
            crypto.Name = remove_quotes(field);

            // Symbol
            getline(ss, field, ',');
            crypto.Symbol = remove_quotes(field);

            // Price
            getline(ss, field, ',');
            crypto.Price = stod(replace_comma_with_dot(remove_quotes(field)));

            // Market Cap
            getline(ss, field, ',');
            crypto.Market_cap = stod(replace_comma_with_dot(remove_quotes(field)));

            // Circulating Supply
            getline(ss, field);
            crypto.Circulating_supply = stod(replace_comma_with_dot(remove_quotes(field)));
        }
        catch (const exception &e)
        {
            cerr << "Ошибка при парсинге строки: " << line << "\nПричина: " << e.what() << endl;
            continue;
        }

        ListNode *newNode = new ListNode{crypto, nullptr};
        if (!list.head)
            list.head = list.tail = newNode;
        else
        {
            list.tail->next = newNode;
            list.tail = newNode;
        }
    }
    file.close();
}

void print(const Cryptocurrency &crypto)
{
    cout << fixed << setprecision(2);
    cout << "Название: " << crypto.Name << "\n"
         << "Символ: " << crypto.Symbol << "\n"
         << "Цена: " << crypto.Price << " USD\n"
         << "Рыночная капитализация: " << crypto.Market_cap << "\n"
         << "Оборот: " << crypto.Circulating_supply << " токенов\n\n";
    cout << defaultfloat;
}

void print_all_cryptos(ListNode *arr[], int size)
{
    cout << "=== Все криптовалюты (" << size << " шт.) ===\n";
    for (int i = 0; i < size; i++)
    {
        print(arr[i]->data);
    }
    cout << "=======================\n";
}

int fill_array(LinkedList &list, ListNode *arr[], int max_count)
{
    int count = 0;
    ListNode *current = list.head;
    while (current && count < max_count)
    {
        arr[count++] = current;
        current = current->next;
    }
    return count;
}

void sort_cryptos(ListNode *arr[], int size)
{
    for (int i = 0; i < size - 1; i++)
    {
        for (int j = 0; j < size - i - 1; j++)
        {
            if (arr[j]->data.Name > arr[j + 1]->data.Name)
                swap(arr[j], arr[j + 1]);
        }
    }
}

void sort_quick_v2(ListNode *arr[], int left, int right)
{
    if (left >= right)
        return;
    string pivot = arr[(left + right) / 2]->data.Name;
    int i = left, j = right;

    while (i <= j)
    {
        while (arr[i]->data.Name < pivot)
        {
            i++;
        }
        while (arr[i]->data.Name > pivot)
        {
            j--;
        }
        if (i <= j)
        {
            swap(arr[i], arr[j]);
            i++;
            j--;
        }
    }
    if (j - left < right - i)
    {
        sort_quick_v2(arr, left, j);
        left = i;
    }
    else
    {
        sort_quick_v2(arr, i, right);
        right = j;
    }
    if (left < right)
        sort_quick_v2(arr, left, right);
}

int binary_search_by_name(ListNode *arr[], int size, const string &target)
{
    int left = 1;
    int right = size - 1;

    while (left < right)
    {
        int mid = left + (right - left) / 2;
        if (arr[mid]->data.Name == target)
            return mid;
        else if (arr[mid]->data.Name < target)
            left = mid + 1;
        else
            right = mid;
    }
    return -1;
}

void free_list(LinkedList &list)
{
    ListNode *current = list.head;
    while (current)
    {
        ListNode *next = current->next;
        delete current;
        current = next;
    }
    list.head = list.tail = nullptr;
}

int main()
{
    setlocale(LC_ALL, "ru");
    LinkedList cryptoList;
    load_from_file(cryptoList);

    ListNode *ptrArr[MAX_CRYPTOS];
    int count = fill_array(cryptoList, ptrArr, MAX_CRYPTOS);
    if (count == 0)
    {
        cerr << "Данные не загружены. Проверьте содержимое файла." << endl;
        return 1;
    }

    sort_cryptos(ptrArr, count);
    // sort_quick_v2(ptrArr, 0, count - 1);

    string query;
    while (true)
    {
        cout << "1. Поиск по названию\n"
             << "2. Вывести все криптовалюты\n"
             << "0. Выход\n"
             << "Выберите действие: ";

        getline(cin, query);

        if (query == "0")
            break;
        else if (query == "1")
        {
            cout << "Введите название криптовалюты: ";
            getline(cin, query);
            int index = binary_search_by_name(ptrArr, count, query);
            if (index != -1)
                print(ptrArr[index]->data);
            else
                cout << "Криптовалюта не найдена!\n";
        }
        else if (query == "2")
            print_all_cryptos(ptrArr, count);
        else
            cout << "Неверный ввод!\n";
    }

    free_list(cryptoList);
    return 0;
}
